### PR TITLE
Add more regression tests

### LINKED ISSUE #39618: Stack overflow with specialization enabled
```rust
#![feature(specialization)]

trait Foo {
    fn foo(&self);
}

trait Bar {
    fn bar(&self);
}

impl<T> Bar for T where T: Foo {
    fn bar(&self) {}
}

impl<T> Foo for T where T: Bar {
    fn foo(&self) {}
}

impl Foo for u64 {}

fn main() {
}

```

Compiling this code results in:
```
thread 'rustc' has overflowed its stack
fatal runtime error: stack overflow
```
There is no backtrace.

If I disable specialization feature, rustc reports an error:
```
error[E0119]: conflicting implementations of trait `Foo` for type `u64`
````

Rustc details:

```
rustc 1.17.0-nightly (0648517fa 2017-02-03)
binary: rustc
commit-hash: 0648517faf1e2cf37c8b6770cbd0180a816ed9a0
commit-date: 2017-02-03
host: x86_64-unknown-linux-gnu
release: 1.17.0-nightly
LLVM version: 3.9
```

### LINKED ISSUE #51798: rustc panicked at 'LocalTableInContext: key not found', libcore/option.rs:960:5
I'm sorry, I haven't reduced this at all, and I'm sure the code is awful, but the error message said you'd appreciate a bug report.

I tried this code:

jeremyBanks/0dmg@ae3b7d261d9616b81d7a1e4cfe11b6e752a3eb0e
using `cargo clean && RUST_BACKTRACE=1 cargo run --verbose`

I expected to see this happen: my program would run, or the compiler would give one of its typical helpful error messages!

Instead, this happened:

```
note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.28.0-nightly (01cc982e9 2018-06-24) running on x86_64-unknown-linux-gnu

note: compiler flags: -C debuginfo=2 -C incremental --crate-type bin

note: some of the compiler flags provided by cargo are hidden
```

## Meta

I am using the Windows Linux Subsystem thing with Debian or Ubuntu, I think.

`rustc --version --verbose`:
```
rustc 1.28.0-nightly (01cc982e9 2018-06-24)
binary: rustc
commit-hash: 01cc982e936120acb0424e41de14e42ba2d88c6f
commit-date: 2018-06-24
host: x86_64-unknown-linux-gnu
release: 1.28.0-nightly
LLVM version: 6.0
```

`cargo --version --verbose`:
```
cargo 1.28.0-nightly (e2348c2db 2018-06-07)
release: 1.28.0
commit-hash: e2348c2db296ce33428933c3ab8786d5f3c54a2e
commit-date: 2018-06-07
```

Backtrace:


```
error: internal compiler error: no type-dependent def for method call                                                                                                                        
  --> src/server.rs:48:17                                                                                                                                                                    
   |                                                                                                                                                                                         
48 |                 display.write_to(&mut encoded_image, image::ImageOutputFormat::PNG);                                                                                                    
   |                 ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^                                                                                                     
                                                                                                                                                                                             
thread 'rustc' panicked at 'LocalTableInContext: key not found', libcore/option.rs:960:5                                                                                                     
stack backtrace:                                                                                                                                                                             
   0:     0x7f6d10abf07e - std::sys::unix::backtrace::tracing::imp::unwind_backtrace::h1b3a1496bdd47d5b                                                                                      
                               at libstd/sys/unix/backtrace/tracing/gcc_s.rs:49                                                                                                              
   1:     0x7f6d10a96046 - std::sys_common::backtrace::print::h2f040662be69463e                                                                                                              
                               at libstd/sys_common/backtrace.rs:71                                                                                                                          
                               at libstd/sys_common/backtrace.rs:59                                                                                                                          
   2:     0x7f6d10ac60ed - std::panicking::default_hook::{{closure}}::h796d09e60146b74e                                                                                                      
                               at libstd/panicking.rs:211                                                                                                                                    
   3:     0x7f6d10ac5e60 - std::panicking::default_hook::h591ad3dc45e922d2                                                                                                                   
                               at libstd/panicking.rs:227                                                                                                                                    
   4:     0x7f6d0d0ed855 - rustc::util::common::panic_hook::h93012055300dea77                                                                                                                
   5:     0x7f6d10ac6873 - std::panicking::rust_panic_with_hook::h704ba6b5b270af4d                                                                                                           
                               at libstd/panicking.rs:515                                                                                                                                    
   6:     0x7f6d10ac65c9 - std::panicking::continue_panic_fmt::hf3634b62bc28487a                                                                                                             
                               at libstd/panicking.rs:426                                                                                                                                    
   7:     0x7f6d10ac6315 - rust_begin_unwind                                                                                                                                                 
                               at libstd/panicking.rs:337                                                                                                                                    
   8:     0x7f6d10b4025b - core::panicking::panic_fmt::hf24f6f6052fa35b2                                                                                                                     
                               at libcore/panicking.rs:92                                                                                                                                    
   9:     0x7f6d10b18bd1 - core::option::expect_failed::hcb17cddc36ef26d5                                                                                                                    
                               at libcore/option.rs:960                                                                                                                                      
  10:     0x7f6d0f62519a - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h9e6d5b384161a649                                             
  11:     0x7f6d0f60820b - rustc::hir::intravisit::walk_local::h9923da8b04f99efa                                                                                                             
  12:     0x7f6d0f607dae - rustc::hir::intravisit::walk_block::h33929fd27434ca57                                                                                                             
  13:     0x7f6d0f62512c - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h9e6d5b384161a649                                             
  14:     0x7f6d0f60820b - rustc::hir::intravisit::walk_local::h9923da8b04f99efa                                                                                                             
  15:     0x7f6d0f607dae - rustc::hir::intravisit::walk_block::h33929fd27434ca57                                                                                                             
  16:     0x7f6d0f62512c - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h9e6d5b384161a649                                             
  17:     0x7f6d0f61671e - rustc::hir::intravisit::walk_expr::he552c42b919e1b81                                                                                                              
  18:     0x7f6d0f62512c - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h9e6d5b384161a649                                             
  19:     0x7f6d0f62512c - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_expr::h9e6d5b384161a649                                             
  20:     0x7f6d0f624d1c - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_nested_body::hdc8c701d5e5bd510                                      
  21:     0x7f6d0f625cd8 - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_impl_item::h072602c658b2a316                                        
  22:     0x7f6d0f617662 - rustc::hir::intravisit::walk_item::h470024412920b632                                                                                                              
  23:     0x7f6d0f6259ed - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item::ha50b33ab6440822d                                             
  24:     0x7f6d0f617976 - rustc::hir::intravisit::walk_item::h470024412920b632                                                                                                              
  25:     0x7f6d0f6259ed - <rustc_privacy::TypePrivacyVisitor<'a, 'tcx> as rustc::hir::intravisit::Visitor<'tcx>>::visit_item::ha50b33ab6440822d                                             
  26:     0x7f6d0f628eb7 - rustc_privacy::privacy_access_levels::hd31f9a283676d277                                                                                                           
  27:     0x7f6d0ce0aa38 - rustc::ty::query::<impl rustc::ty::query::config::QueryAccessors<'tcx> for rustc::ty::query::queries::privacy_access_levels<'tcx>>::compute::hd400623c3414d3e2    
  28:     0x7f6d0cd51f03 - rustc::ty::context::tls::with_context::h03c6fdd2edf55524                                                                                                          
  29:     0x7f6d0cc70498 - rustc::dep_graph::graph::DepGraph::with_task_impl::ha5a5ab6b46058ac3                                                                                              
  30:     0x7f6d0cddee22 - rustc::ty::context::tls::with_related_context::ha15b602b83ee2289                                                                                                  
  31:     0x7f6d0cf02a32 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::force_query_with_job::hbe6ac0c34ea931e3                                            
  32:     0x7f6d0cfc01b4 - rustc::ty::query::plumbing::<impl rustc::ty::context::TyCtxt<'a, 'gcx, 'tcx>>::get_query::h9b8bc4da9a1e7848                                                       
  33:     0x7f6d10e6c082 - rustc::util::common::time::h0fa6f741c87b6466                                                                                                                      
  34:     0x7f6d10f02332 - rustc::ty::context::tls::enter_context::h84146176419ecbc3                                                                                                         
  35:     0x7f6d10eb248a - <std::thread::local::LocalKey<T>>::with::heeb19f7ad79c895c                                                                                                        
  36:     0x7f6d10e19c52 - rustc::ty::context::TyCtxt::create_and_enter::h55f2408bc01b7235                                                                                                   
  37:     0x7f6d10e5d010 - rustc_driver::driver::compile_input::hcefbb4026998c8a0                                                                                                            
  38:     0x7f6d10ee7eab - rustc_driver::run_compiler_with_pool::hf3c65b5bd222c3ca                                                                                                           
  39:     0x7f6d10e4c95b - <scoped_tls::ScopedKey<T>>::set::h8edaec88de07c236                                                                                                                
  40:     0x7f6d10e221cd - syntax::with_globals::hb0aff83358220802                                                                                                                           
  41:     0x7f6d10eba4b2 - std::sys_common::backtrace::__rust_begin_short_backtrace::hbd943a03a0952add                                                                                       
  42:     0x7f6d10ad49e9 - __rust_maybe_catch_panic                                                                                                                                          
                               at libpanic_unwind/lib.rs:105                                                                                                                                 
  43:     0x7f6d10e59fbd - <F as alloc::boxed::FnBox<A>>::call_box::hbe51a9318c82f125                                                                                                        
  44:     0x7f6d10ac561a - std::sys_common::thread::start_thread::h9f808e1cb13674ec                                                                                                          
                               at /checkout/src/liballoc/boxed.rs:650                                                                                                                        
                               at libstd/sys_common/thread.rs:24                                                                                                                             
  45:     0x7f6d10a95015 - std::sys::unix::thread::Thread::new::thread_start::h668a95d388fe6493                                                                                              
                               at libstd/sys/unix/thread.rs:90                                                                                                                               
  46:     0x7f6d0acf76b9 - start_thread                                                                                                                                                      
  47:     0x7f6d1076741c - clone                                                                                                                                                             
  48:                0x0 - <unknown>                                                                                                                                                         
query stack during panic:                                                                                                                                                                    
#0 [privacy_access_levels] privacy access levels                                                                                                                                             
end of query stack                                                                                                                                                                           
error: aborting due to previous error                                                                                                                                                        
```

### LINKED ISSUE #62894: "Aborted (core dumped)" with asm! after "error: expected expression, found `<eof>`"
rustc crashes ungracefully when given the following input (found by [fuzz-rustc](https://github.com/dwrensha/fuzz-rustc)):
```rust
asm!(f(assert_eq!(f/
```

```
$ rustc -vV
rustc 1.38.0-nightly (273f42b59 2019-07-21)
binary: rustc
commit-hash: 273f42b5964c29dda2c5a349dd4655529767b07f
commit-date: 2019-07-21
host: x86_64-unknown-linux-gnu
release: 1.38.0-nightly
LLVM version: 9.0

$ cat crash.rs 
asm!(f(assert_eq!(f/

$ rustc crash.rs 
error: this file contains an un-closed delimiter
 --> crash.rs:1:22
  |
1 | asm!(f(assert_eq!(f/
  |     - -          -   ^
  |     | |          |
  |     | |          un-closed delimiter
  |     | un-closed delimiter
  |     un-closed delimiter

error: macros that expand to items must be delimited with braces or followed by a semicolon
 --> crash.rs:1:5
  |
1 | asm!(f(assert_eq!(f/
  |     ^^^^^^^^^^^^^^^^^
help: change the delimiters to curly braces
  |
1 | asm! {f(assert_eq!(f/}
  |      ^               ^
help: add a semicolon
  |
1 | asm!(f(assert_eq!(f/;
  |                      ^

error[E0658]: use of unstable library feature 'asm': inline assembly is not stable enough for use and is subject to change
 --> crash.rs:1:1
  |
1 | asm!(f(assert_eq!(f/
  | ^^^
  |
  = note: for more information, see https://github.com/rust-lang/rust/issues/29722
  = help: add `#![feature(asm)]` to the crate attributes to enable

error: expected expression, found end of macro arguments
 --> crash.rs:1:21
  |
1 | asm!(f(assert_eq!(f/
  |                     ^ expected expression

Aborted (core dumped)
```

I'm seeing this behavior on stable, beta, and nightly.

### LINKED ISSUE #63952: rustc diverges on large dangling integer slice
I wanted to test our handling of slices that are "too big" (bigger than the largest possible allocation), so I wrote:
```rust
use std::usize;

#[repr(C)]
#[derive(Copy, Clone)]
struct SliceRepr {
    ptr: *const u8,
    len: usize,
}

union SliceTransmute {
    repr: SliceRepr,
    slice: &'static [u8],
}

// bad slice: length too big to even exist anywhere
const SLICE_WAY_TOO_LONG: &[u8] = unsafe { SliceTransmute { repr: SliceRepr { ptr: &42, len: usize::MAX } }.slice};
```
I expected an ICE or so, but I got a timeout. On nightly, here's what a stacktrace looks like:
```
#0  0x00007fe68caece57 in rustc_mir::interpret::visitor::ValueVisitor::walk_value () at src/librustc_mir/interpret/visitor.rs:223
#1  0x00007fe68caec7b1 in rustc_mir::interpret::visitor::ValueVisitor::visit_value () at src/librustc_mir/interpret/visitor.rs:138
#2  rustc_mir::interpret::visitor::ValueVisitor::visit_field () at src/librustc_mir/interpret/visitor.rs:170
#3  rustc_mir::interpret::visitor::ValueVisitor::walk_aggregate () at src/librustc_mir/interpret/visitor.rs:219
#4  0x00007fe68caedb0e in <rustc_mir::interpret::intern::InternVisitor as rustc_mir::interpret::visitor::ValueVisitor<rustc_mir::const_eval::CompileTimeInterpreter>>::visit_aggregate () at src/librustc_mir/interpret/intern.rs:136
#5  rustc_mir::interpret::visitor::ValueVisitor::walk_value () at src/librustc_mir/interpret/visitor.rs:326
#6  0x00007fe68c71b1de in rustc_mir::interpret::visitor::ValueVisitor::visit_value () at src/librustc_mir/interpret/visitor.rs:138
#7  rustc_mir::interpret::intern::intern_const_alloc_recursive () at src/librustc_mir/interpret/intern.rs:284
#8  0x00007fe68c5493b1 in rustc_mir::const_eval::eval_body_using_ecx () at src/librustc_mir/const_eval.rs:161
#9  rustc_mir::const_eval::const_eval_raw_provider::{{closure}} () at src/librustc_mir/const_eval.rs:658
#10 core::result::Result<T,E>::and_then () at /home/r/src/rust/rustc.2/src/libcore/result.rs:691
#11 rustc_mir::const_eval::const_eval_raw_provider () at src/librustc_mir/const_eval.rs:657
```
@oli-obk looks like the interning could do with a fast path for integer slices similar to what validation does? And maybe it could also check that the reference is dereferencable before walking it, that would avoid finding out later that we are going to produce an error anyway.

### LINKED ISSUE #68653: GAT ICE: unelided lifetime in signature
The following ICE's:
```Rust
#![feature(generic_associated_types)]

trait Fun {
    type F<'a: 'a>;
}

impl <T> Fun for T {
    type F<'a> = Self;
}


fn main() {
}
```

<details>
<summary>Backtrace:</summary>

```
error: internal compiler error: unelided lifetime in signature
 --> src/main.rs:4:16
  |
4 |     type F<'a: 'a>;
  |                ^^

thread 'rustc' panicked at 'no errors encountered even though `delay_span_bug` issued', src/librustc_errors/lib.rs:357:17
stack backtrace:
   0: backtrace::backtrace::libunwind::trace
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/libunwind.rs:88
   1: backtrace::backtrace::trace_unsynchronized
             at /cargo/registry/src/github.com-1ecc6299db9ec823/backtrace-0.3.40/src/backtrace/mod.rs:66
   2: std::sys_common::backtrace::_print_fmt
             at src/libstd/sys_common/backtrace.rs:77
   3: <std::sys_common::backtrace::_print::DisplayBacktrace as core::fmt::Display>::fmt
             at src/libstd/sys_common/backtrace.rs:59
   4: core::fmt::write
             at src/libcore/fmt/mod.rs:1052
   5: std::io::Write::write_fmt
             at src/libstd/io/mod.rs:1426
   6: std::sys_common::backtrace::_print
             at src/libstd/sys_common/backtrace.rs:62
   7: std::sys_common::backtrace::print
             at src/libstd/sys_common/backtrace.rs:49
   8: std::panicking::default_hook::{{closure}}
             at src/libstd/panicking.rs:204
   9: std::panicking::default_hook
             at src/libstd/panicking.rs:224
  10: rustc_driver::report_ice
  11: std::panicking::rust_panic_with_hook
             at src/libstd/panicking.rs:476
  12: std::panicking::begin_panic
  13: <rustc_errors::HandlerInner as core::ops::drop::Drop>::drop
  14: core::ptr::drop_in_place
  15: <alloc::rc::Rc<T> as core::ops::drop::Drop>::drop
  16: core::ptr::drop_in_place
  17: rustc_interface::interface::run_compiler_in_existing_thread_pool
  18: scoped_tls::ScopedKey<T>::set
  19: syntax::with_globals
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.

error: internal compiler error: unexpected panic

note: the compiler unexpectedly panicked. this is a bug.

note: we would appreciate a bug report: https://github.com/rust-lang/rust/blob/master/CONTRIBUTING.md#bug-reports

note: rustc 1.42.0-nightly (3761dcd34 2020-01-28) running on x86_64-unknown-linux-gnu

note: compiler flags: -C opt-level=3 -C codegen-units=1 --crate-type bin

note: some of the compiler flags provided by cargo are hidden

query stack during panic:
end of query stack
error: could not compile `playground`.
```
</details>

### PR DESCRIPTION
**Pull request by JohnTitor**
Closes #39618
Closes #51798
Closes #62894
Closes #63952
Closes #68653

r? @Centril

### CODE REVIEW

File: src/test/ui/generic-associated-types/issue-68653.rs [hunk 1]
```diff
@@ -0,0 +1,14 @@
+// check-pass
```
Comment (Centril) [inline src/test/ui/generic-associated-types/issue-68653.rs]:
```suggestion
// A regression test for #68653, which was fixed by #68938.

// check-pass
```

File: src/test/ui/macros/issue-62894.rs [hunk 2]
```diff
@@ -0,0 +1,7 @@
+// Regression test for #62894, shouldn't crash.
+// error-pattern: this file contains an unclosed delimiter
+// error-pattern: expected one of `(`, `[`, or `{`, found keyword `fn`
+
+fn f() { assert_eq!(f(), (), assert_eq!(assert_eq!
```
Comment (Centril) [inline src/test/ui/macros/issue-62894.rs]: Let's move the test into `ui/parser/` as it doesn't seem specific to macro expansion?

File: src/test/ui/specialization/issue-39618.rs [hunk 3]
```diff
@@ -0,0 +1,24 @@
...
+
+impl<T> Bar for T where T: Foo {
+    fn bar(&self) {}
+}
+
+impl<T> Foo for T where T: Bar {
+    fn foo(&self) {}
+}
+
+impl Foo for u64 {}
```
Comment (Centril) [inline src/test/ui/specialization/issue-39618.rs]: This looks suspicions... There's no `default` anywhere, yet logically, we have `u64: Foo`, hence `where T: Foo` is discharged, hence `u64: Bar`, hence `T: Bar` is discharged, and thus `impl<T> Foo for T` applies (and so we have overlap).

cc @jonas-schievink (I thought this had been fixed?)
cc @matthewjasper & @nikomatsakis

**Review by Centril — APPROVED** [review]

**Comment by Centril** [comment]
Oops; didn't mean to approve this. Everything except the specialization looks fine modulo nits.

File: src/test/ui/generic-associated-types/issue-68653.rs (hunk 1 above)
Comment (JohnTitor) [reply src/test/ui/generic-associated-types/issue-68653.rs]: Added.

File: src/test/ui/macros/issue-62894.rs (hunk 2 above)
Comment (JohnTitor) [reply src/test/ui/macros/issue-62894.rs]: Moved.

**Comment by Centril** [comment]
@JohnTitor If you want to get this merged sooner we could also drop the specialization bit and revisit once the discussion has settled. :) If so, r=me with that test removed or a FIXME noting the question added to the specialization test, either at your option.

**Comment by JohnTitor** [comment]
> If so, r=me with that test removed or a FIXME noting the question added to the specialization test, either at your option.

On my local, the test shows SIGSEGV with old nightly mentioned in that issue so I feel it's worth to add this test. I'll add a FIXME note.

**Pull request merged**